File: hopcroft_karp/cpp/hopcroft_karp.cpp
#include <cstring>
#include <fstream>
#include <iostream>
#include <limits>
#include <queue>
#include <vector>

const int INF = std::numeric_limits<int>::max();

class HopcroftKarp
{
public:
    HopcroftKarp(int set_U, int set_V) : U_(set_U), V_(set_V), adjacency_(set_U),
             pair_matching_from_U_(set_U, -1), pair_matching_from_V_(set_V, -1),
             distance_from_free_nodes_(U_ + 1) {}

    void AddEdge(int u, int v)
    {
        adjacency_[u].push_back(v);
    }

    int MaximumMatching()
    {
        int matching = 0;
        while (BFS())
        {
            for (int u = 0; u < U_; ++u)
            {
                if (pair_matching_from_U_[u] == -1 && DFS(u))
                {
                    matching++;
                }
            }
        }

        return matching;
    }

private:
    bool BFS()
    {
        std::queue<int> q;
        for (int u = 0; u < U_; ++u)
        {
            if (pair_matching_from_U_[u] == -1)
            {
                distance_from_free_nodes_[u] = 0;
                q.push(u);
            }
            else
            {
                distance_from_free_nodes_[u] = INF;
            }
        }

        distance_from_free_nodes_[U_] = INF;

        while (!q.empty())
        {
            int u = q.front();
            q.pop();

            if (distance_from_free_nodes_[u] < distance_from_free_nodes_[U_])
            {
                for (int v : adjacency_[u])
                {
                    if (pair_matching_from_V_[v] == -1 || (pair_matching_from_V_[v] != -1
                        && distance_from_free_nodes_[pair_matching_from_V_[v]] == INF))
                    {
                        int nextIndex = (pair_matching_from_V_[v] == -1 ? U_ : pair_matching_from_V_[v]);
                        distance_from_free_nodes_[nextIndex] = distance_from_free_nodes_[u] + 1;
                        q.push(nextIndex);
                    }
                }
            }
        }

        return distance_from_free_nodes_[U_] != INF;
    }

    bool DFS(int u)
    {
        if (u == -1) return true;
        for (int v : adjacency_[u])
        {
            if (pair_matching_from_V_[v] == -1
                || (distance_from_free_nodes_[pair_matching_from_V_[v]] == distance_from_free_nodes_[u] + 1
                && DFS(pair_matching_from_V_[v])))
            {
                pair_matching_from_V_[v] = u;
                pair_matching_from_U_[u] = v;
                return true;
            }
        }

        distance_from_free_nodes_[u] = INF;
        return false;
    }

    int U_, V_;
    std::vector<std::vector<int>> adjacency_;
    std::vector<int> pair_matching_from_U_, pair_matching_from_V_, distance_from_free_nodes_;
};

int main()
{
    std::ios::sync_with_stdio(false);
    std::cin.tie(NULL);
    std::ifstream input_file("hopcroft_karp_input.txt");

    if (!input_file)
    {
        std::cout << "Failed to open input file." << std::endl;
        return 1;
    }

    int nodes_set_U, nodes_set_V, from_u_set, to_v_set;
    input_file >> nodes_set_U >> nodes_set_V;
    HopcroftKarp hk(nodes_set_U, nodes_set_V);

    while (input_file >> from_u_set >> to_v_set)
    {
        hk.AddEdge(from_u_set, to_v_set);
    }

    std::cout << "Maximum Cardinality Matching is " << hk.MaximumMatching() << std::endl;

    return 0;
}

File: hopcroft_karp/python/hopcroft_karp.py
from collections import deque
import sys

INF = sys.maxsize

class HopcroftKarp:
    def __init__(self, set_U, set_V):
        self.U_ = set_U
        self.adjacency_ = [[] for _ in range(set_U)]
        self.pair_matching_from_U_ = [-1] * set_U
        self.pair_matching_from_V_ = [-1] * set_V
        self.distance_from_free_nodes_ = [0] * (set_U + 1)

    def add_edge(self, u, v):
        self.adjacency_[u].append(v)

    def maximum_matching(self):
        matching = 0
        while self.bfs():
            for u in range(self.U_):
                if self.pair_matching_from_U_[u] == -1 and self.dfs(u):
                    matching += 1
        return matching

    def bfs(self):
        q = deque()
        for u in range(self.U_):
            if self.pair_matching_from_U_[u] == -1:
                self.distance_from_free_nodes_[u] = 0
                q.append(u)
            else:
                self.distance_from_free_nodes_[u] = INF

        self.distance_from_free_nodes_[self.U_] = INF

        while q:
            u = q.popleft()
            if self.distance_from_free_nodes_[u] < self.distance_from_free_nodes_[self.U_]:
                for v in self.adjacency_[u]:
                    next_index = self.U_ if self.pair_matching_from_V_[v] == -1 else self.pair_matching_from_V_[v]
                    if self.distance_from_free_nodes_[next_index] == INF:
                        self.distance_from_free_nodes_[next_index] = self.distance_from_free_nodes_[u] + 1
                        q.append(next_index)

        return self.distance_from_free_nodes_[self.U_] != INF

    def dfs(self, u):
        if u == -1:
            return True
        for v in self.adjacency_[u]:
            next_index = self.pair_matching_from_V_[v]
            if next_index == -1 or (self.distance_from_free_nodes_[next_index] == self.distance_from_free_nodes_[u] + 1 and self.dfs(next_index)):
                self.pair_matching_from_V_[v] = u
                self.pair_matching_from_U_[u] = v
                return True

        self.distance_from_free_nodes_[u] = INF
        return False


if __name__ == "__main__":
    try:
        with open("hopcroft_karp_input.txt", "r") as input_file:
            nodes_set_U, nodes_set_V = map(int, input_file.readline().split())
            hk = HopcroftKarp(nodes_set_U, nodes_set_V)

            for line in input_file:
                from_u_set, to_v_set = map(int, line.split())
                hk.add_edge(from_u_set, to_v_set)

    except IOError:
        print("Failed to open input file.")
        exit()

    print("Maximum Cardinality Matching is", hk.maximum_matching())
